source/Condition.h

```diff
@@ -0,0 +1,194 @@
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#ifndef RVALUE_H_
+#define RVALUE_H_
+
+
+#include <cmath>
+#include <limits>
+#include <string>
+#include <type_traits>
+
+
+
+// This class stores either:
+//   1. A condition's value and key
+//   2. A literal value (key is empty)
+//
+// The value (V) is first in the template since (nearly?) all Condition
+// classes will use a std::string key (K). Having a template key type
+// allows storage of scope, wstring, etc. in the future without
+// rewriting this class.
+template < class V, class K = std::string >
+class Condition {
+public:
+	typedef V ValueType;
+	typedef K KeyType;
+
+	constexpr Condition() : value(), key() {}
+	explicit constexpr Condition(const V &value) : value(value), key() {}
+	constexpr Condition(const V &value, const K &key) : value(value), key(key) {}
+	~Condition() {}
+
+	template <class V2, class K2>
+	Condition(const Condition<V2,K2> &other);
+
+	template <class V2, class K2>
+	Condition<V,K> &operator = (const Condition<V2,K2> &other);
+
+	// Update the value from a scope that contains it
+	template<class Getter>
+	const ValueType &UpdateConditions(const Getter &getter);
+
+	// Update the value from a scope that contains it, but use
+	// ValueType() if Validator(weight) is false.
+	template<class Getter, class Validator>
+	const ValueType &UpdateConditions(const Getter &getter, Validator validator);
+
+	// Accessors and mutators
+
+	const ValueType &Value() const { return value; }
+	ValueType &Value() { return value; }
+	const KeyType &Key() const { return key; }
+	KeyType &Key() { return key; }
+
+	// Does this Condition come from the same place as the other one?
+	// If it was a condition, the key must be the same (value doesn't matter)
+	// If it was a literal (no key) then the value must be the same.
+	// If one is literal and the other is conditional, the result is false.
+	bool SameOrigin(const Condition<V,K> &o);
+
+	// Does this originate from a condition?
+	bool HasConditions() const { return !key.empty(); }
+
+	// Does this originate from a literal value (ie. 5.071)?
+	bool IsLiteral() const { return key.empty(); }
+
+	// Floating-point values are false if they're within half the
+	// type's precision of 0 while any other types are passed
+	// through static_cast<bool>
+	explicit operator bool() const;
+
+	// Allow the Condition to be treated as its value in most contexts.
+	operator ValueType() const { return value; }
+
+private:
+	ValueType value;
+	KeyType key;
+};
+
+
+// Allow construction and assignment between Condition types to
+// facilitate type conversion.
```

Comment:
wont this convert it without ppl even realizing it?

Reply:
I will change the constructor to explicit.

Reply:
No, actually, reading through the code, it has to be an implicit conversion. That's because a Condition<T> has to replace arithmetic types in some other template classes. Hence, you have to be able to implicitly convert Conditions just like you can implicitly convert arithmetic types. Its the same reason why there's an operator Value() to implicitly convert to the value. A Condition<T> has to be a drop-in replacement for T.

Reply:
To put it another way, if `Condition<T>` isn't a drop-in replacement for T, then any code that might use conditions will have to be rewritten. The resulting code will be longer and messier than if a Condition could be treated as its value in most expressions.

Reply:
fair let's keep it that way at least for now